pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java

```diff
@@ -0,0 +1,212 @@
+    writer.flush();
+
+    BufferedReader reader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream(), "UTF-8"));
+    return new JSONObject(reader.readLine());
+  }
+
+  private static class ResultPair implements Comparable<ResultPair> {
+    private final String _key;
+    private final long _value;
+
+    public ResultPair(String key, long value) {
+      _key = key;
+      _value = value;
+    }
+
+    @Override
+    public int compareTo(@Nonnull ResultPair o) {
+      return Long.compare(o._value, _value);
+    }
+
+    @Override
+    public String toString() {
+      return _key + ": " + _value;
+    }
+  }
+
+  public static void main(String[] args)
+      throws Exception {
+    // Get all resources in cluster.
+    List<String> resourcesInCluster = HELIX_ADMIN.getResourcesInCluster(CLUSTER_NAME);
+    Collections.sort(resourcesInCluster);
+
+    for (String tableName : resourcesInCluster) {
+
+      // Only process on tables with specific keyword.
+      if (tableName.contains("_non_additive_")) {
+
+        // Get the inverted index config.
+        AbstractTableConfig tableConfig = getTableConfig(tableName);
+        IndexingConfig indexingConfig = tableConfig.getIndexingConfig();
+        List<String> invertedIndexColumns = indexingConfig.getInvertedIndexColumns();
+
+        // Handle null inverted index columns.
+        if (invertedIndexColumns == null) {
+          invertedIndexColumns = new ArrayList<>();
+          indexingConfig.setInvertedIndexColumns(invertedIndexColumns);
+        }
+
+        // Remove empty string.
+        int emptyStringIndex;
+        while ((emptyStringIndex = invertedIndexColumns.indexOf("")) != -1) {
+          invertedIndexColumns.remove(emptyStringIndex);
+        }
+
+        // Only add inverted index if there is no existing ones.
+        if (invertedIndexColumns.isEmpty()) {
+
+          // Table must have a schema to add inverted index.
+          Schema tableSchema = getTableSchema(tableName);
+          if (tableSchema != null) {
+
+            // Skip tables without dimensions.
+            List<String> dimensionNames = tableSchema.getDimensionNames();
+            if (dimensionNames.size() == 0) {
+              continue;
+            }
+
+            // Skip tables without proper time column.
+            String timeColumnName = tableSchema.getTimeColumnName();
+            if (timeColumnName == null || !timeColumnName.equals("Date")) {
+              continue;
+            }
+
+            // Only add inverted index to table larger than a threshold.
+            JSONObject queryResponse = sendQuery("SELECT COUNT(*) FROM " + tableName);
+            long numTotalDocs = queryResponse.getLong("totalDocs");
+            if (numTotalDocs > TABLE_SIZE_THRESHOLD) {
+
+              // Get each dimension's cardinality on one day's data.
+              queryResponse = sendQuery("SELECT Max(Date) FROM " + tableName);
```

Comment:
you might want to get the name of time column as input or derive it from the schema.